astropy/table/table.py

```diff
@@ -3654,6 +3654,9 @@ def copy(self, copy_data=True):
             deepcopied regardless of the value for ``copy_data``.
         """
         out = self.__class__(self, copy=copy_data)
+        if copy_data:
```

Comment:
Hmm, I still think we should only adjust `__deepcopy__` - right now, `Table.copy()` does the same thing as `dict.copy()` (which makes a shallow copy of the values = the data), and I think that's fine.

That will probably remove all errors, but obviously means one has to add extra tests...

But perhaps good to get @taldcroft's opinion first!

Reply:
My mistake! If I understand correctly what you're suggesting is more or less this patch ?
```patch
diff --git a/astropy/table/table.py b/astropy/table/table.py
index 2c6e6c4a40..ca00fbdb69 100644
--- a/astropy/table/table.py
+++ b/astropy/table/table.py
@@ -3663,7 +3663,10 @@ class Table:
         return out
 
     def __deepcopy__(self, memo=None):
-        return self.copy(True)
+        out = self.copy(True)
+        for name in out.colnames:
+            out[name] = deepcopy(out[name])
+        return out
 
     def __copy__(self):
         return self.copy(False)
```
There's probably some duplicated work with this version (but not more than with the *current* state of this PR), but I also note that it also hits the exact same 2 failures that I still don't quite get.

Reply:
Yes, that's what I meant.

And let me if I can figure out the test...

Reply:
I cannot say I necessarily understand it completely, but the following does work:
```
        out = self.copy(copy_data=False)
        for name in out.colnames:
            out.columns.__setitem__(name, deepcopy(self[name]), validated=True)
        return out
```
Note that the strange line is just the very final line of `replace_column` - the rest is not needed since in this case we know the data is OK.  An alternative that works as well is,
```
        out = self.copy(copy_data=False)
        out._replace_cols({name: deepcopy(self[name]) for name in self.colnames})
        return out
```

Maybe @taldcroft can advice on what would be the best procedure!?

Reply:
Thank you very much ! I pushed the first approach just so everyone can see that it indeed passes all tests, but happy to change it again if Tom advises we use a different approach.